src/compiler/checker.ts

```diff
@@ -6638,6 +6691,430 @@ namespace ts {
             }
         }
 
+
+        function checkJsxSelfClosingElement(node: JsxSelfClosingElement) {
+            checkJsxOpeningLikeElement(node);
+            return jsxElementType;
+        }
+
+        function tagNamesAreEquivalent(lhs: EntityName, rhs: EntityName): boolean {
+            if(lhs.kind !== rhs.kind) {
+                return false;
+            }
+            if(lhs.kind === SyntaxKind.Identifier) {
+                return (<Identifier>lhs).text === (<Identifier>rhs).text;
+            }
+            return (<QualifiedName>lhs).right.text === (<QualifiedName>rhs).right.text &&
+                tagNamesAreEquivalent((<QualifiedName>lhs).left, (<QualifiedName>rhs).left);
+        }
+
+        function checkJsxElement(node: JsxElement) {
+            // Check that the closing tag matches
+            if(!tagNamesAreEquivalent(node.openingElement.tagName, node.closingElement.tagName)) {
+                error(node.closingElement, Diagnostics.Expected_corresponding_JSX_closing_tag_for_0, getTextOfNode(node.openingElement.tagName));
+            }
+
+            // Check attributes
+            checkJsxOpeningLikeElement(node.openingElement);
+
+            // Check children
+            let children = node.children;
+            for (var i = 0, n = children.length; i < n; i++) {
+                switch (children[i].kind) {

```

Comment:
I'd just do `let child = children[i]`. Unless this has to be super-optimal, why not just just do a for-of?